google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java

```diff
@@ -0,0 +1,550 @@
+    assertThat(doc.get("foo")).isEqualTo("bar");
+  }
+
+  /**
+   * A tuple class used by {@code #queryWatch}. This class represents an event delivered to the
+   * registered query listener.
+   */
+  private static final class ListenerEvent {
+
+    @Nullable private final QuerySnapshot value;
+    @Nullable private final FirestoreException error;
+
+    ListenerEvent(@Nullable QuerySnapshot value, @Nullable FirestoreException error) {
+      this.value = value;
+      this.error = error;
+    }
+  }
+
+  private static final class EventsCountDownLatch {
+    private final CountDownLatch cdl;
+    private final EnumMap<DocumentChange.Type, CountDownLatch> cdls;
+
+    EventsCountDownLatch(
+        int nonChangeInitial, int addedInitial, int modifiedInitial, int removedInitial) {
+      cdl = new CountDownLatch(nonChangeInitial);
+      cdls = new EnumMap<>(DocumentChange.Type.class);
+      cdls.put(DocumentChange.Type.ADDED, new CountDownLatch(addedInitial));
+      cdls.put(DocumentChange.Type.MODIFIED, new CountDownLatch(modifiedInitial));
+      cdls.put(DocumentChange.Type.REMOVED, new CountDownLatch(removedInitial));
+    }
+
+    void countDown() {
+      cdl.countDown();
+    }
+
+    void countDown(DocumentChange.Type type) {
+      cdls.get(type).countDown();
+    }
+
+    void await() throws InterruptedException {
+      cdl.await(5, TimeUnit.SECONDS);
+    }
+
+    void await(DocumentChange.Type type) throws InterruptedException {
+      cdls.get(type).await(5, TimeUnit.SECONDS);
+    }
+  }
+
+  static class QuerySnapshotEventListener implements EventListener<QuerySnapshot> {
+    final List<ListenerEvent> receivedEvents;
+    final EventsCountDownLatch eventsCountDownLatch;
+
+    QuerySnapshotEventListener(int nonChangeInitial) {
+      this(nonChangeInitial, 0, 0, 0);
+    }
+
+    QuerySnapshotEventListener(
+        int nonChangeInitial, int addedInitial, int modifiedInitial, int removedInitial) {
+      this.receivedEvents = Collections.synchronizedList(new ArrayList<ListenerEvent>());
+      this.eventsCountDownLatch =
+          new EventsCountDownLatch(nonChangeInitial, addedInitial, modifiedInitial, removedInitial);
+    }
+
+    @Override
+    public void onEvent(@Nullable QuerySnapshot value, @Nullable FirestoreException error) {
+      receivedEvents.add(new ListenerEvent(value, error));
+      if (value != null) {
+        List<DocumentChange> documentChanges = value.getDocumentChanges();
+        for (DocumentChange dc : documentChanges) {
+          eventsCountDownLatch.countDown(dc.getType());
+        }
+      }
+      eventsCountDownLatch.countDown();
+    }
+
+    ListenerAssertions assertions() {
+      return new ListenerAssertions(receivedEvents);
+    }
+
+    static final class ListenerAssertions {
+      private static final MapJoiner MAP_JOINER = Joiner.on(",").withKeyValueSeparator("=");
+      final List<ListenerEvent> receivedEvents;
+      private final FluentIterable<ListenerEvent> events;
```

Comment:
Do we need both receivedEvents and events?

Reply:
Nope, leftovers from refactorings. Cleaned up.